### PR TITLE
Remove the two-argument form of AsyncLoader.DoAsync<T>

### PR DESCRIPTION
**Pull request by sharwell**
The two-argument form of `DoAsync<T>` didn't seem to provide substantial advantages over JTF. This pull request converts the few uses to a more direct JTF approach and requires `DoAsync<T>` users pass the third argument.

📝 If accepted, please do not rebase or squash this pull request during the merge.

### CODE REVIEW

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 1]
```diff
@@ -946,12 +946,17 @@ private void UpdateStashCount()
...
             if (AppSettings.ShowStashCount)
             {
-                ThreadHelper.JoinableTaskFactory.RunAsync(() =>
+                ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
                 {
-                    return AsyncLoader.DoAsync(() => Module.GetStashes().Count,
-                        (result) => toolStripSplitStash.Text = string.Format(_stashCount.Text, result,
-                            result != 1 ? _stashPlural.Text : _stashSingular.Text));
-                });
+                    await TaskScheduler.Default.SwitchTo(alwaysYield: true);
```
Comment (jbialobr) [inline GitUI/CommandsDialogs/FormBrowse.cs]: We are on UI thread here, can we drop `alwaysYield: true`?

Comment (sharwell) [reply GitUI/CommandsDialogs/FormBrowse.cs]: ➡️ Yes, `alwaysYield` only has an effect if you are already on the target thread. I'll update both cases for this PR and add assertions to the calling method so the current thread is clear.

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 2]
```diff
@@ -961,6 +968,8 @@ private void UpdateStashCount()
 
         private void CheckForMergeConflicts()
         {
+            ThreadHelper.ThrowIfNotOnUIThread();
```
[the remark this answers is not in the document]
Comment (sharwell) [reply GitUI/CommandsDialogs/FormBrowse.cs]: I'm torn on how to handle this in the long run. In #4728, an empty file was added called **vs-threading.TypesRequiringMainThread.txt**. One could imagine adding the following line to this file:

> System.Windows.Forms.*

However, this produced an incredible burden for development (I think 7,000 new warnings, including many warnings for things that are not actually problematic). For now I'm trying to document the assertion when it's relevant for our code, but we haven't started to leverage its full capability yet - and probably shouldn't until we figure out the usability issues.

[the remark this answers is not in the document]
Comment (sharwell) [reply GitUI/CommandsDialogs/FormBrowse.cs]: > which does not assert we are on the UI thread nor it requests to `alwaysYield`

The use of `alwaysYield` is not characteristic of other cases where vs-threading is used. For us, it serves two purposes:

1. It allowed me to preserve the behavior of code during a transition to vs-threading. By preserving existing behavior, we did not have to examine each case in detail to determine if `alwaysYield` should be specified or not.
2. It hides the latency we have in GitCommands, which does not implement asynchronous commands even though they are run in a separate process. Once GitCommands moves to async code, the need for `alwaysYield` will likely disappear.

[the remark this answers is not in the document]
Comment (jbialobr) [reply GitUI/CommandsDialogs/FormBrowse.cs]: Thank you!

**Pull request merged**